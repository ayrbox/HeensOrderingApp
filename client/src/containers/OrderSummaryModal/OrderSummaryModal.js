import React from 'react';
import PropTypes from 'prop-types';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import Button from '@material-ui/core/Button';
import Grid from '@material-ui/core/Grid';
import pick from 'lodash/pick';

import ConfirmAction from '../../components/ConfirmAction';
import DeliveryAddress from './components/DeliveryAddress';
import OrderTable from './components/OrderTable';
import OrderItems from './components/OrderItems';
import OrderNote from './components/OrderNote';
import Discount from './components/Discount';
import Status from './components/Status';

const OrderSummaryModal = ({
  order,
  setTable,
  setDeliveryAddress,
  addNote,
  setDiscount,
  setStatus,
  processOrder,
  showSummary,
}) => {
  const {
    orderType,
    deliveryAddress,
    subTotal,
    discount,
    orderTotal,
    orderItems,
    tableNo,
    openSummary,
    note,
    status,
  } = order;

  const handleProcessOrder = () => {
    const orderToProcess = {
      ...pick(order, [
        'orderType',
        'deliveryAddress',
        'tableNo',
        'orderItems',
        'subTotal',
        'discount',
        'orderTotal',
        'status',
      ]),
      date: new Date(),
    };
    processOrder(orderToProcess);
  };

  const handleCloseSummary = (e) => {
    e.preventDefault();
    showSummary(false);
  };

  return (
    <Dialog
      open={openSummary}
      aria-labelledby="Order Summary"
      scroll="body"
      maxWidth="xl"
    >
      <DialogContent>
        <Grid
          container
          direction="row"
          justify="flex-start"
          alignItems="flex-start"
        >
          <Grid item xs={6}>
            <OrderItems
              orderItems={orderItems}
              subTotal={subTotal}
              discount={discount}
              total={orderTotal}
            />
          </Grid>
          <Grid item xs={6} style={{ padding: '16px' }}>
            <DeliveryAddress
              type={orderType}
              deliveryAddress={deliveryAddress}
              onChange={setDeliveryAddress}
            />
            <OrderTable
              type={orderType}
              tableNo={tableNo}
              onChange={setTable}
            />
            <OrderNote onChange={addNote} note={note} />
            <Discount onChange={setDiscount} discount={discount} />
            <Status onChange={setStatus} status={status} />
          </Grid>
        </Grid>
      </DialogContent>
      <DialogActions>
        <Button
          onClick={handleCloseSummary}
          size="large"
        >
          Close
        </Button>
        <ConfirmAction
          action={handleProcessOrder}
          message="Please click yes to confirm and process order?"
        >
          <Button
            variant="contained"
            color="primary"
            size="large"
          >
            Confirm Order
          </Button>
        </ConfirmAction>
      </DialogActions>
    </Dialog>
  );
};

OrderSummaryModal.propTypes = {
  order: PropTypes.shape().isRequired,
  setTable: PropTypes.func.isRequired,
  setDeliveryAddress: PropTypes.func.isRequired,
  setDiscount: PropTypes.func.isRequired,
  setStatus: PropTypes.func.isRequired,
  addNote: PropTypes.func.isRequired,
  processOrder: PropTypes.func.isRequired,
  showSummary: PropTypes.func.isRequired,
};

export default OrderSummaryModal;
