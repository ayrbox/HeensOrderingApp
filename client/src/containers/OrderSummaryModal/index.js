import OrderSummary from './connect';

export default OrderSummary;
