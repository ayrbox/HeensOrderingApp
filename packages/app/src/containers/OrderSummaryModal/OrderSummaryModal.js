import React from 'react';
import PropTypes from 'prop-types';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import Button from '@material-ui/core/Button';
import Grid from '@material-ui/core/Grid';
import Typography from '@material-ui/core/Typography';
import pick from 'lodash/pick';
import { withRouter } from 'react-router-dom';

import ConfirmAction from '../../components/ConfirmAction';
import DeliveryAddress from './components/DeliveryAddress';
import OrderTable from './components/OrderTable';
import OrderItems from './components/OrderItems';
import OrderNote from './components/OrderNote';
import Discount from './components/Discount';
import Status from './components/Status';

const OrderSummaryModal = ({
  order,
  setTable,
  setDeliveryAddress,
  addNote,
  setDiscount,
  setStatus,
  processOrder,
  showSummary,
  history,
}) => {
  const {
    orderType,
    deliveryAddress,
    subTotal,
    discount,
    orderTotal,
    orderItems,
    tableNo,
    openSummary,
    note,
    status,
    requestInProgress,
    requestSuccess,
  } = order;

  const handleProcessOrder = () => {
    const orderToProcess = {
      ...pick(order, [
        'orderType',
        'deliveryAddress',
        'tableNo',
        'orderItems',
        'subTotal',
        'discount',
        'orderTotal',
        'status',
      ]),
      date: new Date(),
    };
    processOrder(orderToProcess);
  };

  const handleCloseSummary = (e) => {
    e.preventDefault();
    showSummary(false);
  };

  return (
    <Dialog
      open={openSummary}
      aria-labelledby="Order Summary"
      scroll="body"
      maxWidth="xl"
      onClose={handleCloseSummary}
    >
      <DialogContent>
        <Grid
          container
          direction="row"
          justify="flex-start"
          alignItems="flex-start"
        >
          <Grid item xs={6}>
            <OrderItems
              orderItems={orderItems}
              subTotal={subTotal}
              discount={discount}
              total={orderTotal}
            />
          </Grid>
          <Grid item xs={6} style={{ padding: '16px' }}>
            <DeliveryAddress
              type={orderType}
              deliveryAddress={deliveryAddress}
              onChange={setDeliveryAddress}
            />
            <OrderTable
              type={orderType}
              tableNo={tableNo}
              onChange={setTable}
            />
            <OrderNote onChange={addNote} note={note} />
            <Discount onChange={setDiscount} discount={discount} />
            <Status onChange={setStatus} status={status} />
          </Grid>
        </Grid>
      </DialogContent>
      <DialogActions>
        {!requestInProgress && !requestSuccess && (
          <>
            <Button
              onClick={handleCloseSummary}
              size="large"
            >
              Close
            </Button>
            <ConfirmAction
              action={handleProcessOrder}
              message="Please click yes to confirm and process order?"
            >
              <Button
                variant="contained"
                color="primary"
                size="large"
              >
                Confirm Order
              </Button>
            </ConfirmAction>
          </>
        )}
        {!requestInProgress && requestSuccess && (
          <>
            <Typography
              variant="body1"
              style={{ color: 'green' }}
            >
              Order saved.
            </Typography>
            <Button
              size="large"
              onClick={() => history.push('/orders')}
            >
              Go to orders
            </Button>
          </>
        )}
      </DialogActions>
    </Dialog>
  );
};

OrderSummaryModal.propTypes = {
  order: PropTypes.shape().isRequired,
  setTable: PropTypes.func.isRequired,
  setDeliveryAddress: PropTypes.func.isRequired,
  setDiscount: PropTypes.func.isRequired,
  setStatus: PropTypes.func.isRequired,
  addNote: PropTypes.func.isRequired,
  processOrder: PropTypes.func.isRequired,
  showSummary: PropTypes.func.isRequired,
  history: PropTypes.shape().isRequired,
};

export default withRouter(OrderSummaryModal);
