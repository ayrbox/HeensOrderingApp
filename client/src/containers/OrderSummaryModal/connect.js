import { connect } from 'react-redux';
import OrderSummaryModal from './OrderSummaryModal';

import { processOrder } from '../../actions/newOrderActions';

const mapState = state => ({
  order: state.newOrder,
});

export default connect(mapState, {
  processOrder,
})(OrderSummaryModal);
