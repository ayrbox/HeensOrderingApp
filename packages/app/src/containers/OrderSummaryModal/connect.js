import { connect } from 'react-redux';
import OrderSummaryModal from './OrderSummaryModal';

import {
  setTable,
  setDeliveryAddress,
  setDiscount,
  setStatus,
  addNote,
  processOrder,
  showSummary,
} from '../../store/newOrder';

const mapState = state => ({
  order: state.newOrder,
});

export default connect(mapState, {
  setTable,
  setDeliveryAddress,
  addNote,
  setDiscount,
  setStatus,
  processOrder,
  showSummary,
})(OrderSummaryModal);
