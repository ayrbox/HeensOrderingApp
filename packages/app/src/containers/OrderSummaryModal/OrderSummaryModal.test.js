import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import {
  render,
  fireEvent,
  cleanup,
  waitForElement,
} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import OrderSummaryModal from './OrderSummaryModal';

const sampleOrder = {
  requestInProgress: false,
  requestSuccess: false,
  openNewOrderPane: false,
  orderType: 'table',
  orderItems: [
    {
      _id: '23982A',
      name: 'Crispy Seaweed干貝',
      description: 'Crispy Seaweed',
      price: 4.9,
      menuOptions: [],
      itemTotal: 4.9,
    },
    {
      _id: '23982B',
      name: 'Sesame Prawn on Toast下多士',
      description: 'Sesame Prawn on Toast',
      price: 5.5,
      menuOptions: [],
      itemTotal: 5.5,
    },
    {
      _id: '23982C',
      name: 'Heen\'s Mixed Hors D\'oeuvres并',
      description: 'Mixed Hors D\'oeuvres (For 2 Persons or More)',
      price: 7.9,
      menuOptions: [],
      itemTotal: 7.9,
    },
    {
      _id: '23982D',
      name: 'Vegetarian Mini Spring Rolls齐卷',
      description: 'Vegetarian Mini Spring Rolls',
      price: 4.9,
      menuOptions: [],
      itemTotal: 4.9,
    },
  ],
  subTotal: 36,
  discount: 0,
  orderTotal: 36,
  note: '',
  openMenuModal: false,
  openSummary: true,
  status: 'ordered',
};


const mockProcessOrder = jest.fn();
const mockShowSummary = jest.fn();
const mockResetOrder = jest.fn();

describe('<OrderTable />', () => {
  afterEach(cleanup);

  it('renders correctly', async () => {
    const { container, getByTestId } = render(
      <Router>
        <OrderSummaryModal
          order={sampleOrder}
          setTable={jest.fn()}
          setDeliveryAddress={jest.fn()}
          setDiscount={jest.fn()}
          setStatus={jest.fn()}
          addNote={jest.fn()}
          processOrder={mockProcessOrder}
          showSummary={mockShowSummary}
          resetOrder={mockResetOrder}
        />
      </Router>,
    );

    // debug();

    fireEvent.click(getByTestId('close-order-button'));
    expect(mockShowSummary).toHaveBeenCalled();

    // triggers confirm modal and then triggers process func
    fireEvent.click(getByTestId('confirm-order-button'));
    const [, yesButton] = await waitForElement(() => [
      getByTestId(/confirm-modal/),
      getByTestId(/yes-button/),
    ]);
    fireEvent.click(yesButton);
    expect(mockProcessOrder).toHaveBeenCalled();


    // test snapshot for the component
    expect(container).toMatchSnapshot();
  });
});
