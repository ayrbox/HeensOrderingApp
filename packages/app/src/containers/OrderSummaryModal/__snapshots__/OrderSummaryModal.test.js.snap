// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OrderTable /> renders correctly 1`] = `
<div
  aria-hidden="true"
/>
`;
